core/src/main/java/ai/timefold/solver/core/impl/move/director/VariableChangeRecordingScoreDirector.java

```diff
@@ -203,4 +203,20 @@ public void recordListAssignment(ListVariableDescriptor<Solution_> variableDescr
             variableChanges.add(new ListVariableAfterAssignmentAction<>(element, variableDescriptor));
         }
     }
+
+    /**
+     * Record a before list assigment.
+     * Used when moves with nested phases create a new solution but do not record events before the list changes,
+     * such as ruin and recreate.
+     * The before event must be manually recorded,
+     * or the original list will not be restored during the rollback action.
+     *
+     * @param variableDescriptor never null
+     * @param entity never null
+     * @param list never null and may be empty
+     */
+    public void recordBeforeListVariableChanged(ListVariableDescriptor<Solution_> variableDescriptor, Object entity,
```

Comment:
Two questions:


1. If this is `recordBeforeListVariableChanged`, does it mean that `recordListAssignment()` should instead be `recordAfterListVariableChanged`?
2.  We typically call the usual before/after methods on the score director, except in this case, where we have a whole new bunch of methods. It is confusing. It looks as is these methods are glued to the recording score director "from the side". They look like they don't belong. I wonder... does the score director need refactoring in order for these methods to fit more naturally?